include/grpc/event_engine/event_engine.h

```diff
@@ -208,42 +245,59 @@ class EventEngine {
     /// Called with the result of a TXT record lookup
     using LookupTXTCallback = std::function<void(absl::Status, std::string)>;
 
-    virtual ~DNSResolver() = 0;
+    virtual ~DNSResolver() = default;
 
-    // TODO(hork): define status codes for the callback
-    /// Asynchronously resolve an address. \a default_port may be a non-numeric
-    /// named service port, and will only be used if \a address does not already
-    /// contain a port component.
+    /// Asynchronously resolve an address.
+    ///
+    /// \a default_port may be a non-numeric named service port, and will only
+    /// be used if \a address does not already contain a port component.
+    ///
+    /// \a on_resolve will be called exactly once. Implementations are expected
```

Comment:
Suggest wording this similarly to `Endpoint::Read()` above:

"""
The \a on_resolve callback will be called when the lookup is complete.  Implementations should pass the appropriate statuses to the callback. For example, callbacks might expect to receive DEADLINE_EXCEEDED when the deadline is exceeded or CANCELLED if the lookup was cancelled.
"""

Reply:
I avoided saying "when the lookup is complete" because even though it's true (e.g., a lookup failure counts as a completion), that wording on its own might be interpreted as "when the lookup has completed successfully" which is only part of the story.

How about changing the first sentence to:

"""
The \a on_resolve callback will be called when the lookup has completed successfully or failed.
"""

Reply:
I don't actually think that "when the lookup is complete" implies anything about the success or failure of the lookup, especially since the rest of the paragraph talks about the status being passed to the callback.  But if you're concerned about that, how about this wording:

"""
When the lookup is complete, the \a on_resolve callback will be invoked with a status indicating the success or failure of the lookup.  Implementations should pass the appropriate statuses to the callback. For example, callbacks might expect to receive DEADLINE_EXCEEDED when the deadline is exceeded or CANCELLED if the lookup was cancelled.
"""

Hmmm... Would it make sense to change the callback arguments from `absl::Status` and `std::vector<ResolvedAddress>` to a single argument of type `absl::StatusOr<std::vector<ResolvedAddress>>`?   That way, the API makes it clearer that it will either fail or get a result -- i.e., there will never be a case where both arguments are relevant at the same time.

Reply:
That language works, thanks. And yes, StatusOr would make sense for the Lookup callbacks and OnConnectCallback. I've made those changes.